tests/base/functions_10_move.cc

```diff
@@ -0,0 +1,149 @@
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+
+// Test InterpolatedTensorProductGridData
+//
+// Like the _10 test, but using the move constructor that moves the
+// data from the initializing tables into the
+// InterpolatedTensorProductGridData object.
+
+#include <deal.II/base/function_lib.h>
+
+#include "../tests.h"
+
+// now interpolate the function x*y*z onto points. note that this function is
+// (bi/tri)linear and so we can later know what the correct value is that the
+// function should provide
+Table<1, double>
+fill(const std::array<std::vector<double>, 1> &coordinates)
+{
+  Table<1, double> data(coordinates[0].size());
+  for (unsigned int i = 0; i < coordinates[0].size(); ++i)
+    data[i] = coordinates[0][i];
+  return data;
+}
+
+Table<2, double>
+fill(const std::array<std::vector<double>, 2> &coordinates)
+{
+  Table<2, double> data(coordinates[0].size(), coordinates[1].size());
+  for (unsigned int i = 0; i < coordinates[0].size(); ++i)
+    for (unsigned int j = 0; j < coordinates[1].size(); ++j)
+      data[i][j] = coordinates[0][i] * coordinates[1][j];
+  return data;
+}
+
+Table<3, double>
+fill(const std::array<std::vector<double>, 3> &coordinates)
+{
+  Table<3, double> data(coordinates[0].size(),
+                        coordinates[1].size(),
+                        coordinates[2].size());
+  for (unsigned int i = 0; i < coordinates[0].size(); ++i)
+    for (unsigned int j = 0; j < coordinates[1].size(); ++j)
+      for (unsigned int k = 0; k < coordinates[2].size(); ++k)
+        data[i][j][k] =
+          coordinates[0][i] * coordinates[1][j] * coordinates[2][k];
+  return data;
+}
+
+
+template <int dim>
+void
+check()
+{
+  // have coordinate arrays that span an interval starting at d+1
+  // d+5 nonuniform intervals
+  std::array<std::vector<double>, dim> coordinates;
+  for (unsigned int d = 0; d < dim; ++d)
+    for (unsigned int i = 0; i < d + 5; ++i)
+      coordinates[d].push_back(d + 1 + 1. * i * i);
+
+  const std::array<std::vector<double>, dim> coordinates_backup = coordinates;
+
+  Table<dim, double> data = fill(coordinates);
+
+  Functions::InterpolatedTensorProductGridData<dim> f(std::move(coordinates),
+                                                      std::move(data));
+
+  // Verify tha the data has really been moved
+  for (unsigned int d = 0; d < dim; ++d)
+    Assert(data.size()[d] == 0, ExcInternalError());
+  for (unsigned int d = 0; d < dim; ++d)
+    Assert(coordinates[d].size() == 0, ExcInternalError());
```

Comment:
Hmmm. Can we really rely on accessing the moved-from variables to yield something predictable?

Reply:
We certainly can for this class, where we know what happens. I might not recommend this in user code, but in the test suite, I think this ok.

The standard says in general that moved-from objects are in an undefined state, but in practice, this state must be well-enough defined that one can run the destructor on these objects. In the current case, I cannot see that a `std::vector` that has been moved from could reasonably yield anything other than zero for `size()`.